snuba/clickhouse/native.py

```diff
@@ -222,15 +222,13 @@ def execute(
         query: ClickhouseQuery,
         # TODO: move Clickhouse specific arguments into DictClickhouseQuery
         settings: Optional[Mapping[str, str]] = None,
```

Comment:
It seems to me that settings is now a concept of the abstract reader that is not treated by the DB specific reader implementation as an opaque bag of properties for the DB anymore.
I think we should take an explicit direction on this topic (not for this PR, but asking to see if we are on the same page). Either we allow the callsite to pass an opaque bag of properties to the DB, but then we could not do what this PR is doing by extracting one specific property. Or we give settings a more specific and abstract structure (maybe a data class with the specific fields we can pass to the reader) and let the Reader implementations decide how to make sense of them and transform that class into something DB specific. 
What do you think? I prefer the second.  
Again, not for this PR, but agreeing on the direction will be useful.

Reply:
This is somewhat described here: https://github.com/getsentry/snuba/pull/889#issuecomment-618063054

> 2\. This seems to me like more evidence that query settings are part of the physical query model. Each of these steps uses data from some combination of the request query, query (just for hashing the contents for key generation), and settings to decide how it should be executed. If `ClickhouseQuery` is a subtype of `PhysicalQuery` ([this thread](https://github.com/getsentry/snuba/pull/880#discussion_r412571686)) and the query settings also are consolidated into that class, this would greatly reduce the number of values being passed around here.

In general I think the `Reader` interface should be simplified:

```diff
 class Reader(ABC, Generic[TQuery]):
     @abstractmethod
-    def execute(
-        self,
-        query: TQuery,
-        settings: Optional[Mapping[str, str]] = None,
-        query_id: Optional[str] = None,
-        with_totals: bool = False,
-    ) -> Result:
+    def execute(self, query: TQuery) -> Result:
         """Execute a query."""
         raise NotImplementedError
```

My reasoning:

1. settings should be encapsulated on `TQuery` as they will vary in semantics and structure based on different database engines
2. `with_totals` seems as though it should be a property of the query itself

I think this also aligns with this comment, if I understand it correctly (the `TQuery` that already exists is equivalent to the query type described): https://github.com/getsentry/snuba/pull/880#discussion_r415955384

Reply:
yeah having settings encapsulated in the query will go in the same direction.

Reply:
Cool, all these pieces seem like they're starting to come together in a way that makes a bit more sense.